php/class-customize-concurrency.php

```diff
@@ -14,4 +14,355 @@
+	 *
+	 * @action customize_controls_enqueue_scripts
+	 */
+	public function customize_controls_enqueue_scripts() {
+		wp_enqueue_script( $this->plugin->slug );
+		wp_enqueue_style( $this->plugin->slug );
+	}
+
+	/**
+	 * Add script specifically for the preview pane.
+	 *
+	 * @action customize_preview_init
+	 */
+	function customize_preview_init() {
+		wp_enqueue_script( "{$this->plugin->slug}-preview" );
+	}
+
+	/**
+	 * Send timestamp for when this session started and output template.
+	 */
+	public function customize_controls_print_footer_scripts() {
+		$data = array(
+			'session_start_timestamp' => strtotime( current_time( 'mysql', true ) ),
+		);
+
+		printf( '<script>var _customizeConcurrency = %s;</script>', wp_json_encode( $data ) );
+
+		?>
+		<script type="text/html" id="tmpl-customize-concurrency-notifications">
+			<ul>
+				<# _.each( data.notifications, function( notification ) { #>
+					<li class="notice notice-{{ notification.type || 'info' }} {{ data.altNotice ? 'notice-alt' : '' }} notice-concurrency_conflict" data-code="{{ notification.code }}" data-type="{{ notification.type }}">
+						<# if ( /concurrency_conflict/.test( notification.code ) ) { #>
+							<button class="button concurrency-conflict-override" type="button" data-tooltip="Reject Change"><span class="dashicons dashicons-thumbs-down"></span></button>
+							<button class="button concurrency-conflict-accept" type="button" data-tooltip="Accept Change"><span class="dashicons dashicons-thumbs-up"></span></button>
+							Conflict due to concurrent update by {{ notification.data.user }}.
+							<p><b>Change:</b> <i>"{{ notification.data.their_value }}"</i></p>
+						<# } #>
+					</li>
+				<# } ); #>
+			</ul>
+		</script>
+		<?php
+
+	}
+
+	/**
+	 * Prevent sanitize_title() on the post_name.
+	 *
+	 * @see wp_insert_post()
+	 * @filter wp_insert_post_data
+	 *
+	 * @param array $data    An array of slashed post data.
+	 * @param array $postarr An array of sanitized, but otherwise unmodified post data.
+	 * @return array
+	 */
+	public function preserve_inserted_post_name( $data, $postarr ) {
+		if ( self::POST_TYPE === $data['post_type'] && isset( $postarr['post_name'] ) ) {
+			$data['post_name'] = wp_slash( $postarr['post_name'] );
+		}
+		return $data;
+	}
+
+	/**
+	 * Set up conflict checking within snapshots.
+	 */
+	public function customize_snapshot_save_before( $snapshot, \WP_Customize_Manager $wp_customize ) {
+		$saved_settings = $snapshot->data();
+		$this->sanitize_conflicts( $saved_settings, $wp_customize );
+	}
+
+	/**
+	 * Set up conflict checking without snapshots.
+	 *
+	 * @todo detect snapshot save and do not validate against live changes
+	 */
+	public function customize_register( \WP_Customize_Manager $wp_customize ) {
+		$post_values = $wp_customize->unsanitized_post_values();
+		$saved_settings = $this->get_saved_settings( array_keys( $post_values ) );
+		$this->sanitize_conflicts( $saved_settings, $wp_customize );
```

Comment:
@lgedeon regarding the todo:

> detect snapshot save and do not validate against live changes

Will this not be addressed by the introduction of the new `customize_snapshot_save_before` action? But now that you have `customize_snapshot_save_before`, will that mean that this line calling `sanitize_conflicts` can be removed and the problem will be solved since it only needs to be called in the `customize_snapshot_save_before` action? Sorry I haven't been able to follow up your questions very well.